### PR TITLE
Adapter priority

### PR DESCRIPTION
**Pull request by Rsedaikin**
More flexible adapter selection in Metal and DirectX rendering via properties (skiko.metal.gpu.priority, skiko.derectx.gpu.priority).

### CODE REVIEW

File: skiko/src/jvmMain/cpp/windows/directXRedrawer.cc [hunk 1]
```diff
@@ -220,40 +225,75 @@ extern "C"
...
+                IDXGIAdapter1 *pAdapter = nullptr;
+                if (!SUCCEEDED(factory6->EnumAdapterByGpuPreference(adapterIndex, adapterPriority, IID_PPV_ARGS(&pAdapter))))
+                {
+                    break;
+                }
+
+                DXGI_ADAPTER_DESC1 desc;
+                pAdapter->GetDesc1(&desc);
+                std::wstring currentAdapterName(desc.Description);
+                fwprintf(stderr, L"Adapter #%d: %s  %d\n", adapterIndex, currentAdapterName.c_str(), adapterPriority);
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/windows/directXRedrawer.cc:247]: We shall not print anything by default, please remove the printf.

File: skiko/src/jvmMain/cpp/windows/directXRedrawer.cc [hunk 2]
```diff
@@ -208,6 +209,10 @@ extern "C"
     {
         DXGI_ADAPTER_DESC1 desc;
         hardwareAdapter->GetDesc1(&desc);
+        if (desc.Flags & DXGI_ADAPTER_FLAG_SOFTWARE)
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/windows/directXRedrawer.cc:212]: Better `if ((desc.Flags & DXGI_ADAPTER_FLAG_SOFTWARE) != 0)`

File: skiko/src/jvmMain/cpp/windows/directXRedrawer.cc [hunk 3]
```diff
@@ -220,40 +225,75 @@ extern "C"
...
+                if (DXGI_ERROR_NOT_FOUND == pFactory->EnumAdapters1(adapterIndex, &pAdapter))
                 {
-                    continue;
+                    break;
                 }
-                *ppAdapter = pAdapter;
-                return true;
+                if (SUCCEEDED(D3D12CreateDevice(pAdapter, D3D_FEATURE_LEVEL_11_0, _uuidof(ID3D12Device), nullptr)))
+                {
+                    if (isBlacklisted(pAdapter))
```
Comment (olonho) [inline skiko/src/jvmMain/cpp/windows/directXRedrawer.cc:274]: Shalln't we release blacklisted adapter?

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/redrawer/Direct3DRedrawer.kt [hunk 4]
```diff
@@ -64,13 +64,23 @@ internal class Direct3DRedrawer(
...
 
     fun finishFrame(device: Long, context: Long, surface: Long) {
         finishFrame(device, context, surface, properties.isVsyncEnabled)
     }
 
-    external fun createDirectXDevice(windowHandle: Long): Long
+    fun getAdapterPriority(): Int {
+        val adapterPriority = System.getProperty("skiko.directx.gpu.priority")
+        return when (adapterPriority) {
+            "auto" -> 0
```
Comment (olonho) [inline skiko/src/jvmMain/kotlin/org/jetbrains/skiko/redrawer/Direct3DRedrawer.kt:76]: Maybe enum?

File: skiko/src/jvmMain/cpp/windows/directXRedrawer.cc (hunk 1 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/windows/directXRedrawer.cc]: Sure. Done.

File: skiko/src/jvmMain/cpp/windows/directXRedrawer.cc (hunk 2 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/windows/directXRedrawer.cc]: Done.

File: skiko/src/jvmMain/cpp/windows/directXRedrawer.cc (hunk 3 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/cpp/windows/directXRedrawer.cc:269]: Yes, we shall. Done.

File: skiko/src/jvmMain/kotlin/org/jetbrains/skiko/redrawer/Direct3DRedrawer.kt (hunk 4 above)
Comment (Rsedaikin) [reply skiko/src/jvmMain/kotlin/org/jetbrains/skiko/redrawer/Direct3DRedrawer.kt]: Done.

File: skiko/src/jvmMain/objectiveC/macos/MetalRedrawer.mm [hunk 5]
```diff
@@ -156,7 +166,7 @@ BOOL isUsingIntegratedGPU() {
 }
 
 JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_redrawer_MetalRedrawer_createMetalDevice(
-    JNIEnv *env, jobject redrawer, jlong platformInfoPtr)
+    JNIEnv *env, jobject redrawer, jint adpapterPriority, jlong platformInfoPtr)
```
Comment (olonho) [inline skiko/src/jvmMain/objectiveC/macos/MetalRedrawer.mm:169]: `adapterPriority`

**Review by olonho — APPROVED** [review]
Please fix typo and feel free to commit

**Pull request merged**